_posts/2024-03-20-strimzi-kraft-migration.md

```diff
@@ -0,0 +1,265 @@
+    strimzi.io/node-pools: enabled
+# ...
+```  
+
+You can do that by annotating the custom resource.
+
+```shell
+kubectl annotate kafka my-cluster strimzi.io/kraft="rollback" --overwrite
+```
+
+Because of the rollback, the cluster metadata state moves back to `KRaftDualWriting`, waiting for the user to finalize the revert.
+
+```shell
+NAME         DESIRED KAFKA REPLICAS   DESIRED ZK REPLICAS   READY   METADATA STATE   WARNINGS
+...
+my-cluster   3                        3                             ZooKeeper        True
+my-cluster   3                        3                     True    KRaftMigration   
+my-cluster   3                        3                     True    KRaftDualWriting   
+my-cluster   3                        3                     True    KRaftPostMigration   
+my-cluster   3                        3                     True    KRaftPostMigration   True
+my-cluster   3                        3                     True    KRaftPostMigration   True
+my-cluster   3                        3                     True    KRaftDualWriting     
+my-cluster   3                        3                     True    KRaftDualWriting     True
+my-cluster   3                        3                             KRaftDualWriting     
+```
+
+After that, you have to delete the `KafkaNodePool` hosting the KRaft controllers and then apply the `disabled` value annotation.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/kraft: disabled # updated from "rollback" to "disabled" to finalize the rollback to ZooKeeper
+    strimzi.io/node-pools: enabled
+# ...
+```
+
+You can do that by annotating the custom resource.
+
+```shell
+kubectl annotate kafka my-cluster strimzi.io/kraft="disabled" --overwrite
+```
+
+The annotation change triggers the Strimzi operator to run the following steps for you:
+
+* automatically delete the `/controller` znode on ZooKeeper to allow the brokers to elect the new controller among them in place of the previous KRaft controller;
+* reconfigure brokers without connection details to the KRaft controller quorum anymore but using ZooKeeper again and roll them;
+
+Finally, the cluster metadata state comes back to be `ZooKeeper`.
+
+```shell
+NAME         DESIRED KAFKA REPLICAS   DESIRED ZK REPLICAS   READY   METADATA STATE   WARNINGS
+...
+my-cluster   3                        3                             KRaftDualWriting     
+my-cluster   3                        3                             KRaftDualWriting     
+my-cluster   3                        3                     True    ZooKeeper            True
+```
+
+### Conclusion
+
+While the manual migration process could be considered long and complex if your cluster is running on bare-metal or virtual machines, it is very easy if the cluster is running on Kubernetes instead because of the Strimzi support for it.
```

Comment:
```suggestion
While the manual migration process could be considered long and complex if your cluster is running on bare-metal or virtual machines, it is very easy if the cluster is running on Kubernetes because of the Strimzi support for it.
```